src/jacobi.c

```diff
@@ -0,0 +1,219 @@
+    vec_size sizeSquared = size * size;
+
+    // Iteration starts at 2 since we assume that the first element off diagonal is the max at that point
+    for (vec_size i = 2; i < sizeSquared; ++i) {
+        vec_size currentRow = i / size;
+        vec_size currentCol = i % size;
+        real_number elem = real_abs(matrix[coordToIndex(currentRow, currentCol, size)]);
+
+        if (currentRow != currentCol && elem > maxElement) {
+            currentMaxRow = currentRow;
+            currentMaxCol = currentCol;
+            maxElement = elem;
+        }
+    }
+
+    *maxRow = currentMaxRow;
+    *maxCol = currentMaxCol;
+}
+
+
+/**
+ * Multiplies 2 matrices together using this equation : output = firstMatrix x secondMatrix
+ * @param firstMatrix first matrix to multiply of size m * n
+ * @param secondMatrix second matrix to multiply of size n * p
+ * @param size array of size 3 containing m, n and p
+ * @param output matrix resulting of the matrix multiplication of the given matrices
+ * @param transposeFirstMatrix whether to transpose the first matrix when multiplying the matrices
+ */
+void jacobiMatrixMultiply(real_number* firstMatrix, real_number* secondMatrix, vec_size size[3], real_number* output, vec_size transposeFirstMatrix) {
+    
+    vec_size m = size[0];
+    vec_size n = size[1];
+    vec_size p = size[2];
+    for (vec_size i = 0; i < m; ++i) {
+        for (vec_size j = 0; j < p; ++j) {
+            real_number sum = 0;
+            for (vec_size k = 0; k < n; ++k) {
+                sum += (transposeFirstMatrix ? firstMatrix[coordToIndex(k, i, n)] : firstMatrix[coordToIndex(i, k, n)]) * secondMatrix[coordToIndex(k, j, n)];
+            }
+            output[coordToIndex(i, j, p)] = sum;
+        }
+    }
+}
+
+/**
+ * Creates a jacobi rotation matrix (1 on the diagonal and the value of cos(angle) and sin(angle) at row p and q)
+ * @param row row where to put the c and s values
+ * @param col column where to put the c and s values
+ * @param output output matrix (must be square)
+ * @param size size of the output matrix
+*/
+void jacobiCreateRotationMatrix(real_number* input, vec_size row, vec_size col, real_number* output, vec_size size) {
+
+    real_number aqq = input[coordToIndex(col, col, size)];
+    real_number app = input[coordToIndex(row, row, size)];
+    real_number apq = input[coordToIndex(row, col, size)];
+
+    real_number tau = (aqq - app) / (2.0 * apq);
+    real_number t1 = -tau - sqrt(1 + tau * tau);
+    real_number t2 = -tau + sqrt(1 + tau * tau);
+    real_number t = real_abs(t1) > real_abs(t2) ? t1 : t2;
+
+    real_number c = 1.0 / (sqrt(1 + t * t));
+    real_number s = c * t;
+
+    vec_size cIndex1 = coordToIndex(row, row, size);
+    vec_size cIndex2 = coordToIndex(col, col, size);
+    vec_size sIndex1 = coordToIndex(row, col, size);
+    vec_size sIndex2 = coordToIndex(col, row, size);
+    vec_size sizeSquared = size * size;
+
+    for (vec_size i = 0; i < sizeSquared; ++i) {
```

Comment:
Would it be possible to make this into a switch?

Reply:
I cannot make this into a switch, because the values compared in a switch statement must be known at compile time. In my case, they are not